### ISSUE TITLE
Recursive datastructures

### ISSUE THREAD

**Issue by mlegenhausen** [gcanti/tcomb#71]
How can I define recursive datastructures?

Example:

```js
var Category = strict({
  name: Str,
  children: list(Category)
});
```

This throws an error, cause at definition time `Category` is undefined.

**Comment by timoxley** [comment]
@gcanti This is indeed a bit hacky, can you see a better solution?

**Comment by gcanti** [comment]
It's hacky indeed! I can't think of a better or simpler solution though.

**Comment by mlegenhausen** [comment]
Have Idea what about a Lazy type? Which needs to be overwritten (can be checked on first object creation). Which would be the other extreme:

```js
var Category = struct({
  children: Lazy // or LateBind, Overwrite?
});

new Category(); // throws error

// Can be used
Category = Category.extend({ children: list(Category) });
```

or a lazy evaluated type which binds on first creation:

```js
var Category = struct({
  children: list(lazy(() => Category))
}, 'Category');
```

Other idea, would be a self type, but which would only work on direct dependencies:

```js
var Category = struct({
  children: Self // Which references Category
});
```

**Comment by timoxley** [comment]
The latter seems simpler to me.

**Comment by gcanti** [comment]
A new proposal here https://github.com/gcanti/tcomb/pull/157

**Comment by gcanti** [comment]
Closed via a7c9cd3
